include/cilantro/core/covariance.hpp

```diff
@@ -0,0 +1,160 @@
+            return (*this)(points, nn.begin(), nn.end(), mean, cov);
+        }
+    };
+
+    template <typename ScalarT, ptrdiff_t EigenDim, typename CovarianceT=Covariance<ScalarT, EigenDim>>
+    class MinimumCovarianceDeterminant {
+    public:
+        EIGEN_MAKE_ALIGNED_OPERATOR_NEW
+
+        MinimumCovarianceDeterminant() = default;
+
+        inline bool operator()(const ConstVectorSetMatrixMap<ScalarT,EigenDim> &points, Vector<ScalarT,EigenDim>& mean, Eigen::Matrix<ScalarT,EigenDim,EigenDim>& cov) const {
+            Neighborhood<ScalarT> neighbors(points.cols());
+            int count = 0;
+            std::generate(neighbors.begin(), neighbors.end(), [&count]() mutable { return Neighbor<ScalarT>(count++, 0); });
+            return (*this)(points, neighbors, mean, cov);
+        }
+
+        template <typename NeighborhoodResultIteratorT>
+        inline bool operator()(const ConstVectorSetMatrixMap<ScalarT,EigenDim> &points, NeighborhoodResultIteratorT begin, NeighborhoodResultIteratorT end, Vector<ScalarT,EigenDim>& mean, Eigen::Matrix<ScalarT,EigenDim,EigenDim>& cov) const {
+            const int first_idx = begin->index;
+
+            const size_t size = std::distance(begin, end);
+            if (size < points.rows()) return false;
+            if (size == points.rows()) return compute_mean_and_covariance_(points, begin, end, mean, cov);
+
+            random_selector<> random {};
+
+            using NeighborT = typename NeighborhoodResultIteratorT::value_type;
+
+            std::vector<NeighborT> subset(points.rows());
+            size_t h = static_cast<size_t>(std::ceil(outlier_rate_ * (size + EigenDim + 1)));
+            if (h > size) h = size - 1;
+            Vector<ScalarT,EigenDim> best_mean;
+            Eigen::Matrix<ScalarT,EigenDim,EigenDim> best_cov;
+            ScalarT best_determinant = std::numeric_limits<ScalarT>::max();
+            for (int j = 0; j < num_trials_; ++j) {
+                std::generate(subset.begin(), subset.end(), [&begin, &end, &random]() { return *random(begin, end); });
+                compute_mean_and_covariance_(points, subset.begin(), subset.end(), mean, cov);
+                for (int l = 0; l < num_refinements_; ++l) {
+                    mahalanobisDistance(points, begin, end, mean, cov.inverse());
+                    std::partial_sort(begin, begin + h, end, typename NeighborT::ValueLessComparator());
+                    compute_mean_and_covariance_(points, begin, begin + h, mean, cov);
+                }
+                ScalarT determinant = cov.determinant();
+                if (determinant < best_determinant) {
+                    best_determinant = determinant;
+                    best_cov = cov;
+                    best_mean = mean;
+                }
+            }
+            mean = best_mean;
+            cov = best_cov;
```

Comment:
I think this may still need a finalization procedure to recompute the mean and covariance based on all the inliers and not a partial sort of the inliers.